gateway-js/src/supergraphManagers/IntrospectAndCompose/index.ts

```diff
@@ -0,0 +1,164 @@
+import {
+  composeAndValidate,
+  compositionHasErrors,
+  ServiceDefinition,
+} from '@apollo/federation';
+import { Logger } from 'apollo-server-types';
+import { HeadersInit } from 'node-fetch';
+import resolvable from '@josephg/resolvable';
+import {
+  ServiceEndpointDefinition,
+  SupergraphSdlUpdateFunction,
+  SubgraphHealthCheckFunction,
+} from '../..';
+import {
+  loadServicesFromRemoteEndpoint,
+  Service,
+} from './loadServicesFromRemoteEndpoint';
+import { SupergraphManager, SupergraphSdlHookOptions } from '../../config';
+
+export interface IntrospectAndComposeOptions {
+  subgraphs: ServiceEndpointDefinition[];
+  introspectionHeaders?:
+    | HeadersInit
+    | ((
+        service: ServiceEndpointDefinition,
+      ) => Promise<HeadersInit> | HeadersInit);
+  pollIntervalInMs?: number;
+  logger?: Logger;
+  subgraphHealthCheck?: boolean;
+}
+
+type State =
+  | { phase: 'initialized' }
+  | { phase: 'polling'; pollingPromise?: Promise<void> }
+  | { phase: 'stopped' };
+
+export class IntrospectAndCompose implements SupergraphManager {
+  private config: IntrospectAndComposeOptions;
+  private update?: SupergraphSdlUpdateFunction;
+  private healthCheck?: SubgraphHealthCheckFunction;
+  private subgraphs?: Service[];
+  private serviceSdlCache: Map<string, string> = new Map();
+  private timerRef: NodeJS.Timeout | null = null;
+  private state: State;
+
+  constructor(options: IntrospectAndComposeOptions) {
+    this.config = options;
+    this.state = { phase: 'initialized' };
+  }
+
+  public async initialize({ update, getDataSource, healthCheck }: SupergraphSdlHookOptions) {
+    this.update = update;
+
+    if (this.config.subgraphHealthCheck) {
+      this.healthCheck = healthCheck;
+    }
+
+    this.subgraphs = this.config.subgraphs.map((subgraph) => ({
+      ...subgraph,
+      dataSource: getDataSource(subgraph),
+    }));
+
+    let initialSupergraphSdl: string | null = null;
+    try {
+      initialSupergraphSdl = await this.updateSupergraphSdl();
+    } catch (e) {
+      this.logUpdateFailure(e);
+      throw e;
+    }
+
+    // Start polling after we resolve the first supergraph
+    if (this.config.pollIntervalInMs) {
+      this.beginPolling();
+    }
+
+    return {
+      // on init, this supergraphSdl should never actually be `null`.
+      // `this.updateSupergraphSdl()` will only return null if the schema hasn't
+      // changed over the course of an _update_.
+      supergraphSdl: initialSupergraphSdl!,
+      cleanup: async () => {
+        if (this.state.phase === 'polling') {
+          await this.state.pollingPromise;
+        }
+        this.state = { phase: 'stopped' };
+        if (this.timerRef) {
+          this.timerRef.unref();
+          clearInterval(this.timerRef);
```

Comment:
Should be clearTimer, not clearInterval. (Using timer rather than interval is good because you don't have to worry about overlapping calls.)

Reply:
2a63f07e